### PR TITLE
Dshare 938 f 2024 7639 possibly invalid swap rate due to using unreliable market price

### PR DESCRIPTION
**Pull request by jaketimothy**
In reviewing the `convertToShares` and `convertToAssets` methods and adding testing to those view methods, additional math errors were corrected.

Fixes resulted in bytecode too large for our opt runs and additional changes were made to simplify the code.

### CODE REVIEW

File: src/common/OracleLib.sol [hunk 1]
```diff
@@ -20,4 +20,26 @@ library OracleLib {
...
         return mulDiv(paymentTokenQuantity, decimalMult, assetTokenQuantity);
     }
+
+    function applyPriceAssetToPayment(uint256 assetTokenQuantity, uint256 price, uint8 paymentTokenDecimals)
+        internal
+        pure
+        returns (uint256)
+    {
+        // Adjusts the payment token decimals to match the asset token decimals (18) and removes 18 decimals from the result.
+        uint256 decimalMult = 10 ** (36 - paymentTokenDecimals);
```
Comment (joshualyguessennd) [inline src/common/OracleLib.sol:30]: I'd like to understand if we specifically chose the 36 approach for code size optimization? If not, we could make the decimal handling more explicit by separating the adjustments

File: src/common/OracleLib.sol [hunk 2]
```diff
@@ -20,4 +20,26 @@ library OracleLib {
...
+        return mulDiv(assetTokenQuantity, price, decimalMult);
+    }
+
+    function applyPricePaymentToAsset(uint256 paymentTokenQuantity, uint256 price, uint8 paymentTokenDecimals)
+        internal
+        pure
+        returns (uint256)
+    {
+        // Adjusts the payment token decimals to match the asset token decimals (18) and removes 18 decimals from the result.
+        uint256 decimalMult = 10 ** (36 - paymentTokenDecimals);
```
Comment (joshualyguessennd) [inline src/common/OracleLib.sol:41]: same here

File: src/common/OracleLib.sol (hunk 1 above)
Comment (jaketimothy) [reply src/common/OracleLib.sol:30]: This math is a consequence of `price` having 18 decimals

Comment (jaketimothy) [reply src/common/OracleLib.sol:30]: And asset token also having 18 decimals

**Review by joshualyguessennd — APPROVED** [review]

**Pull request merged**